File: packages/engine/src/cpus/6502/instructions/dex.js
import { CycleType } from '..'

/**
 * DEcrement X
 *
 * X,Z,N = X-1
 *
 * Affects Flags: N Z
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       DEX           $CA  1   2
 */
export function* dex() {
  if (--this.x < 0) this.x = 0xff
  this.zero = this.x === 0
  this.negative = (this.x & 0x80) > 0
  yield CycleType.GET
}

File: packages/engine/src/cpus/6502/instructions/dex.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// DEcrement X
//
// X,Z,N = X-1
//
// Affects Flags: N Z
//
// MODE          SYNTAX        HEX LEN TIM
// Implied       DEX           $CA  1   2

pub fn dex(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    self.x -%= 1;
    self.z = self.x == 0;
    self.n = (self.x & 0x80) > 0;
}
